### ISSUE TITLE
[no-useless-constructor] allow generics

### ISSUE THREAD

**Issue by tao-cumplido** [typescript-eslint/typescript-eslint#1131]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-useless-constructor": "error"
  }
}
```

```TS
class A {
    constructor(readonly value: any) {}
}

class B<T> extends A {
    get value(): T {
        return super.value;
    }

    constructor(value: T) {
        super(value);
    }
}
```

**Expected Result**
No error, or maybe an option to allow it.

**Actual Result**
Useless constructor.eslint(@typescript-eslint/no-useless-constructor)

**Additional Info**
Functionally the constructor in `B` is useless, but without it I'd have to explicitly type the generic `new B<number>(1)`.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.4.0` |
| `@typescript-eslint/parser`        | `2.4.0` |
| `TypeScript`                       | `3.5.3` |
| `ESLint`                           | `6.5.1` |
| `node`                             | `12.11.0` |
| `npm`                              | `X.Y.Z` |

**Comment by bradzacher** [comment]
I think that this is an edge case which is really hard for us to detect correctly.
Happy for someone to have a go at it.

**Comment by mdouglass** [comment]
I just ran into a related case and figured I'd drop it here rather than opening another issue (though feel free to redirect me):

```
class BaseContext {
  constructor(parent?: IContext) { }
}
class Context {
  constructor(parent: IContext) { super(parent) }
}
```

Similar to the original issue, the derived class constructor is "useless" but it does enforce that undefined values are not allowed to be provided to the derived class constructor so it can't be safely removed. 

Could this rule (by default or by option) not trigger if the parameter types are different from the base class constructor's parameter types? That seems like it would catch both this case and the original one.

**Comment by bradzacher** [comment]
> Could this rule (by default or by option) not trigger if the parameter types are different from the base class constructor's parameter types?

No. This would require type information, which the rule does not use.

**Comment by mdouglass** [comment]
That makes sense.

I’m new to the internals here - is it possible to make a rule that optionally uses type information?

**Comment by bradzacher** [comment]
it is possible, but it's not something we do.

It creates a confusing experience for end users (i.e. turn on this one option and your lint will break because now this specific rule will require a different parser config).

**Comment by Delagen** [comment]
I think it has false positive in case of Angular
For example

class BaseContext {
  constructor(service: IService) { }
}
class Context {
  constructor(@Inject(ConcreteService) service: IConcreteService) { super(service) }
}

**Comment by bradzacher** [comment]
@Delagen - that is independent of this issue.
Please file a new issue with that case.

**Comment by Kenya-West** [comment]
Same here

```
export class InputWithSliderComponent extends AbstractFormControlComponent<number> {
    constructor(injector: Injector) {
        super(injector);
    }
}
```

![image](https://user-images.githubusercontent.com/12271104/126424866-8c115116-fa14-488e-a3bb-034c3a074f63.png)

**Comment by JoshuaKGoldberg** [comment]
> type information

I'm going to go ahead and close this for the same reasons as #948. See https://github.com/typescript-eslint/typescript-eslint/issues/948#issuecomment-1326539046 for more reasoning. Thanks all!
